src/status_im/ui2/screens/chat/composer/input.cljs

```diff
@@ -168,12 +174,130 @@
+                                                  (let [content (string/trim paste-content)
+                                                        new-text (calculate-input-text params content)]
+                                                    (update-input-text params new-text)))]
+                                   (react/get-from-clipboard callback)))
+
+                        :biu (fn [{:keys [first-level text-input-handle menu-items selection-start selection-end]}]
+                               (let [show-menu #(.startActionMode (selectable-text-input-manager) text-input-handle)]
+                                 (reset! first-level false)
+                                 (reset! menu-items second-level-menu-items)
+                                 ;to avoid something discusting like this https://lightrun.com/answers/facebook-react-native-textinput-controlled-selection-broken-on-both-ios-and-android
+                                 ;use native invoke instead! do not use setNativeProps! e.g. (.setNativeProps ^js text-input (clj->js {:selection {:start selection-start :end selection-end}}))
+                                 (.setSelection (selectable-text-input-manager) text-input-handle selection-start selection-end)
+                                 ;set a delay to wait RNSelectableTextInputViewManager#_menuItems get updated, need figure out a better way
+                                 (js/setTimeout show-menu 200)))})
+
+(def first-level-menu-items (map i18n/label (keys first-level-menus)))
+
+(defn reset-to-first-level-menu [first-level menu-items]
+  (reset! first-level true)
+  (reset! menu-items first-level-menu-items))
+
+(defn append-markdown-char [{:keys [first-level menu-items content] :as params} wrap-chars]
+  (let [content (str wrap-chars content wrap-chars)
+        new-text (calculate-input-text params content)]
+    (update-input-text params new-text)
+    (reset-to-first-level-menu first-level menu-items)))
+
+(def second-level-menus {:bold #(append-markdown-char % "**")
+
+                         :italic #(append-markdown-char % "*")
+
+                         :strikethrough #(append-markdown-char % "~~")})
+
+(def second-level-menu-items (map i18n/label (keys second-level-menus)))
+
+(defn on-menu-item-touched [{:keys [first-level event-type] :as params}]
+  (let [menus (if @first-level first-level-menus second-level-menus)
+        menu-item-key (nth (keys menus) event-type)
+        action (get menus menu-item-key)]
+    (action params)))
+
+(defn selectable-text-input [chat-id {:keys [style] :as props} children]
+  (let [text-input-ref (reagent/atom nil)
+        menu-items (reagent/atom first-level-menu-items)
+        first-level (reagent/atom true)]
+    (reagent/create-class
+     {:component-did-mount (fn [self]
+                             (when @text-input-ref
+                               (let [self-handle (rn/find-node-handle self)
```

Comment:
We usually use `this`

Reply:
Can this be a functional component instead of a class component?

Reply:
Hi @OmarBasem , i would prefer keep the implementation of class component to functional component ATM as u know there would cause new issue (can't type in the message input) after using functional component :) , @flexsurfer agreed